src/ctx.cpp

```diff
@@ -1656,8 +1639,124 @@ static int lArrayVectorWidth(llvm::Type *t) {
     return (int)arrayType->getNumElements();
 }
 
+std::string lChooseDivisionName(llvm::Instruction::BinaryOps inst, llvm::Type *type) {
+    switch (inst) {
+    case llvm::Instruction::SDiv:
+        if (type == LLVMTypes::Int8VectorType) {
+            return builtin::__sdiv_i8;
+        } else if (type == LLVMTypes::Int16VectorType) {
+            return builtin::__sdiv_i16;
+        } else if (type == LLVMTypes::Int32VectorType) {
+            return builtin::__sdiv_i32;
+        } else if (type == LLVMTypes::Int64VectorType) {
+            return builtin::__sdiv_i64;
+        }
+        break;
+    case llvm::Instruction::UDiv:
+        if (type == LLVMTypes::Int8VectorType) {
+            return builtin::__udiv_i8;
+        } else if (type == LLVMTypes::Int16VectorType) {
+            return builtin::__udiv_i16;
+        } else if (type == LLVMTypes::Int32VectorType) {
+            return builtin::__udiv_i32;
+        } else if (type == LLVMTypes::Int64VectorType) {
+            return builtin::__udiv_i64;
+        }
+        break;
+    case llvm::Instruction::SRem:
+        if (type == LLVMTypes::Int8VectorType) {
+            return builtin::__srem_i8;
+        } else if (type == LLVMTypes::Int16VectorType) {
+            return builtin::__srem_i16;
+        } else if (type == LLVMTypes::Int32VectorType) {
+            return builtin::__srem_i32;
+        } else if (type == LLVMTypes::Int64VectorType) {
+            return builtin::__srem_i64;
+        }
+        break;
+    case llvm::Instruction::URem:
+        if (type == LLVMTypes::Int8VectorType) {
+            return builtin::__urem_i8;
+        } else if (type == LLVMTypes::Int16VectorType) {
+            return builtin::__urem_i16;
+        } else if (type == LLVMTypes::Int32VectorType) {
+            return builtin::__urem_i32;
+        } else if (type == LLVMTypes::Int64VectorType) {
+            return builtin::__urem_i64;
+        }
+        break;
+    default:
+        break;
+    }
+    Assert(false && "Invalid division instruction");
+    return "";
+}
+
+// To avoid exception of "division by zero" in the case of vector division, we
+// can't generate a simple vector division for the whole vector and then mask
+// it. We need to perform division only for the active lanes. It is done by
+// calling builtin __idiv function. If all lanes are active, we can still use
+// the instction to divide the entire vector, which can be useful for
+// optimization when we divide compile time constants, e.g, programIndex
+// similar to the tests/lit-tests/2468.ispc.
+llvm::Value *FunctionEmitContext::VectorIntDivision(llvm::Instruction::BinaryOps inst, llvm::Value *v0, llvm::Value *v1,
+                                                    const llvm::Twine &name) {
+    // When handling vector division, we need to create a branching structure:
+    // If fullmask is all ON, use direct SDiv instruction
+    // Otherwise, use the __idiv intrinsic which handles masked division
```

Comment:
Could you remind me if we perform reciprocal optimization somewhere later? Like where the divisor has the same value across all vector lanes, we can just compute rcp and multiply.

Reply:
I have fixed the case when a divisor is a constant, and added the corresponding test `1415-2.ispc`, and benchmarks (fastdiv_mask).